### PR TITLE
Collapse/Expand Virtual Nodes

### LINKED ISSUE #12: Collapsing/expanding nodes
Not supported by default in dash-cytoscape: https://github.com/plotly/dash-cytoscape/issues/104
May have to do some hacking (e.g. set child nodes to invisible, or remove child nodes entirely)

### PR DESCRIPTION
**Pull request by riverchen99**
closes #12

### CODE REVIEW

File: protos/graph_structures.proto [hunk 1]
```diff
@@ -61,6 +59,16 @@ message Client {
     // ---
 }
 
+message VirtualNode {
+    optional string name = 1;
+    repeated string child_names = 2;
+    optional bool collapsed = 3;
+    optional NodeType node_type = 4; // should always be NODETYPE_VIRTUAL, used for duck typing
```
Comment (pwat2) [inline protos/graph_structures.proto:66]: set default as well?

File: ujt/state.py [hunk 2]
```diff
@@ -66,43 +67,194 @@ def read_local_data() -> Tuple[Dict[str, Node], Dict[str, Client]]:
...
+            parent_virtual_node_map[node_data["ujt_id"]] = virtual_node_name
+        else:
+            node_frontier.append(node_name_message_map[node_data["ujt_id"]])
+
+    while node_frontier:
+        node = node_frontier.popleft()
+        virtual_node_child_names.add(node.name)
+        parent_virtual_node_map[node.name] = virtual_node_name
+        for child_name in node.child_names:
+            node_frontier.append(node_name_message_map[child_name])
```
Comment (riverchen99) [inline ujt/state.py:210]: Here's the code to create a new virtual node. We keep track of a virtual node's children. This parent/child relationship is subtly different from those between actual nodes, as this relationship isn't reflected in the Node proto itself. It's only held on the virtual node's end. Moreover, if a virtual node X contains Service1, Service1.Endpoint1 and Service1.Endpoint2, X's children are all the aforementioned nodes. (They are collapsed).

File: ujt/transformers.py [hunk 3]
```diff
@@ -28,38 +35,142 @@ def apply_highlighted_edge_class_to_elements(elements, user_journey_name):
...
+            if new_source is not None:
+                element["data"]["source"] = new_source
+            if new_target is not None:
+                element["data"]["target"] = new_target
+
+            element["data"][
+                "id"] = f"{element['data']['source']}/{element['data']['target']}"
+
+            if element["data"]["source"] != element["data"]["target"]:
+                new_elements.append(element)
```
Comment (riverchen99) [inline ujt/transformers.py:119]: Here we transform cytoscape elements to account for the currently collapsed virtual nodes. We remove the nodes hidden within a virtual node. Moreover, we adjust the affected edges that originally connected to nodes within the virtual node, to point to the virtual node itself.

File: ujt/transformers.py [hunk 4]
```diff
@@ -28,38 +35,142 @@ def apply_highlighted_edge_class_to_elements(elements, user_journey_name):
...
+    be called on the element to update the aforementioned properties.
+    However, Dash Cytoscape doesn't expose this functionality.
+    By providing a new ID, we can avoid this restriction.
+
+    Args:
+        elements: a list of Cytoscape elements
+
+    Returns:
+        A list of Cytoscape elements with an UUID appended to their ID fields. 
+    """
```
Comment (riverchen99) [inline ujt/transformers.py:165]: This is somewhat of a hack, but it works...

File: ujt/callbacks.py [hunk 5]
```diff
@@ -40,54 +32,137 @@
...
+    compute_status.reset_client_statses(client_name_message_map)
+    compute_status.reset_node_statuses(virtual_node_map)
+
+    # combine the two maps of nodes into one dictionary
+    # use duck typing -- is this pythonic or a hack?
+    all_nodes_map = {**node_name_message_map, **virtual_node_map}  #type: ignore
+    compute_status.compute_statuses(
+        all_nodes_map,
+        client_name_message_map,
+    )
```
Comment (riverchen99) [inline ujt/callbacks.py:116]: Interested to hear your thoughts about this.
Virtual Nodes have children and their status should be computed based on the status of their children.
However, virtual nodes don't have dependencies. I take advantage of python's dick typing here to reuse the existing status computation code for virtual nodes. I think this approach is more elegant, but can also see why we would want to have a specific function to compute status for virtual nodes, even if some of the code is repeated.

(The same situation arises in the callback to create the node info panel.)

File: protos/graph_structures.proto (hunk 1 above)
Comment (riverchen99) [reply protos/graph_structures.proto:66]: WIll do.

File: ujt/callbacks.py (hunk 5 above)
Comment (pwat2) [reply ujt/callbacks.py:116]: this is probably better than having two separate functions.

**Review by pwat2 — APPROVED** [review]

**Review by KenArthur — APPROVED** [review]

**Pull request merged**